File: packages/sdk-ts/src/gateway.ts
/**
 * The Warrant Gateway interface, seen from the adapters.
 *
 * This is the **only** dependency the tools have. The MCP server, the SDK client
 * and the Vercel AI adapter know nothing else: no HTTP, no base64, no
 * `PAYMENT-REQUIRED` headers. All the transport machinery stays behind this
 * boundary, which makes the four tools testable without a network.
 *
 * Trade-off accepted: the Gateway must translate its HTTP 402 into an explicit
 * `{ status: 'payment-required' }` rather than throw an exception. A payment
 * requirement is not an error — it is a normal step of the protocol, and
 * modelling it as an error pushes every caller into doing flow control with
 * `catch`.
 */

import type {
  ActionCategory,
  ActionSpec,
  Address,
  CheckResult,
  ConditionSpec,
  Hex,
  WarrantStatus,
} from 'warrant-core'
import type { PaymentPayload, PaymentRequired, SettlementResponse } from './x402.js'

// ─────────────────────────────────────────────────────────────────────────────
// quote_risk
// ─────────────────────────────────────────────────────────────────────────────

export interface QuoteRequest {
  actionSpec: ActionSpec
  /**
   * Beneficiary of any slash. Optional at quote time: it does not enter the
   * price, only the post-condition built when it is supplied.
   */
  beneficiary?: Address
}

export interface QuoteResult {
  /** Derived from the calldata by the Classifier. Never declared by the agent. */
  category: ActionCategory
  /** Bond in atomic units of the bonding asset (USDC, 6 decimals). */
  bond: string
  riskBps: number
  /** Notional derived from the decoded arguments — published to make the price legible. */
  notionalUSD: string
  conditionSpec: ConditionSpec
  /** One-sentence explanation of the price arrived at. */
  rationale: string
}

// ─────────────────────────────────────────────────────────────────────────────
// request_warrant
// ─────────────────────────────────────────────────────────────────────────────

export interface WarrantRequest {
  actionSpec: ActionSpec
  beneficiary: Address
}

export interface WarrantOpened {
  warrantId: Hex
  /** KeeperHub execution identifier. */
  executionId: string
  conditionHash: Hex
  actionHash: Hex
  /** Unix timestamp in seconds. */
  expiry: number
  bond?: string
  fundingRef?: Hex
}

/**
 * Two outcomes, a single shape. `payment-required` carries the object to be
 * copied verbatim into the MCP tool result; `opened` carries the warrant and, if
 * the x402 rail was taken, the settlement to place back into `_meta`.
 */
export type RequestWarrantResult =
  | { status: 'payment-required'; paymentRequired: PaymentRequired }
  | { status: 'opened'; warrant: WarrantOpened; settlement?: SettlementResponse }

// ─────────────────────────────────────────────────────────────────────────────
// get_warrant / list_warrants
// ─────────────────────────────────────────────────────────────────────────────

export interface WarrantVerdict {
  verdict: 'honored' | 'slashed'
  evaluatedAtBlock: string
  /** RPC used for the evaluation — published to make the verdict replayable. */
  rpcUrl: string
  txHash?: Hex
  blockNumber?: string
  gasUsed?: string
  outcome?: string
  settlementTx?: Hex
  reputationTx?: Hex
}

export interface WarrantView {
  warrantId: Hex
  agent: Address
  beneficiary: Address
  bond: string
  conditionHash: Hex
  actionHash: Hex
  fundingRef?: Hex
  expiry: number
  openedAt: number
  status: WarrantStatus
  category?: ActionCategory
  executionId?: string
  actionSpec?: ActionSpec
  conditionSpec?: ConditionSpec
  /** Present only once the warrant is settled. */
  verdict?: WarrantVerdict
  /**
   * One result per check, including the ones that pass — a verdict that shows
   * only the failed check is not auditable (docs/04). Empty array as long as the
   * warrant is unsettled.
   */
  checks: CheckResult[]
}

export interface ListWarrantsQuery {
  agent: Address
  status?: 'open' | 'honored' | 'slashed' | 'reclaimed'
  category?: ActionCategory
  /** Unix timestamps in seconds, against `openedAt`. */
  since?: number
  until?: number
  limit?: number
  cursor?: string
}

export interface WarrantStats {
  total: number
  open: number
  honored: number
  slashed: number
  reclaimed: number
  /** Sums in atomic units. */
  totalBonded: string
  totalSlashed: string
  /** Share of settled warrants that were honored, in basis points. */
  honorRateBps: number
}

export interface ListWarrantsResult {
  warrants: WarrantView[]
  stats: WarrantStats
  nextCursor?: string
}

// ─────────────────────────────────────────────────────────────────────────────

/**
 * The four calls the Gateway must serve. Nothing more.
 *
 * `requestWarrant` is the only one that takes an optional payment: it is also
 * the only paid tool, and that asymmetry ought to be visible in the signature.
 */
export interface GatewayClient {
  quote(req: QuoteRequest): Promise<QuoteResult>
  requestWarrant(req: WarrantRequest, payment?: PaymentPayload): Promise<RequestWarrantResult>
  /** `null` — and not an exception — when the warrant does not exist. */
  getWarrant(warrantId: Hex): Promise<WarrantView | null>
  listWarrants(query: ListWarrantsQuery): Promise<ListWarrantsResult>
}

File: packages/sdk-py/src/warrant_sdk/gateway.py
"""The Gateway interface, as the adapters see it.

This is the **only** dependency of the four tools. Neither the tools nor the
framework adapters know about HTTP, base64 or ``PAYMENT-REQUIRED`` headers, which
is what makes them testable without a network.

The counterpart is deliberate: a Gateway implementation must translate its HTTP
402 into an explicit :class:`RequestWarrantResult` with
``status="payment-required"`` rather than raising. A payment requirement is not an
error — it is a normal step of the protocol, and modelling it as an exception
pushes every caller into flow control by ``except``.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Protocol

__all__ = ["GatewayClient", "RequestWarrantResult"]


@dataclass(frozen=True)
class RequestWarrantResult:
    """Two outcomes, one shape.

    Attributes:
        status: ``"payment-required"`` or ``"opened"``.
        payment_required: The x402 challenge, to be handed to a signer as-is.
        warrant: The opened warrant — ``warrantId``, ``executionId``,
            ``conditionHash``, ``actionHash``, ``expiry``.
        settlement: The x402 settlement receipt, when the payment rail was used.
    """

    status: str
    payment_required: dict[str, Any] | None = None
    warrant: dict[str, Any] | None = None
    settlement: dict[str, Any] | None = None


class GatewayClient(Protocol):
    """The four calls a Gateway must serve. Nothing more.

    ``request_warrant`` is the only one that takes a payment, because it is the
    only paid tool — an asymmetry that should be visible in the signature rather
    than hidden in a flag.
    """

    def quote(self, request: dict[str, Any]) -> dict[str, Any]:
        """Price an action. Free, and commits nothing."""
        ...

    def request_warrant(
        self,
        request: dict[str, Any],
        payment: dict[str, Any] | None = None,
        *,
        warrant_nonce: str | None = None,
    ) -> RequestWarrantResult:
        """Open a bonded warrant.

        Args:
            request: ``{"actionSpec": …, "beneficiary": …}``.
            payment: An x402 ``PaymentPayload``. Omit it to receive the challenge.
            warrant_nonce: The warrant nonce announced in the 402 challenge. It
                must travel back unchanged with the payment: it determines the
                warrant id, which enters the ``termsHash`` the EIP-3009
                authorization carries as its nonce. Guessing a new one produces a
                different ``termsHash`` than the one signed, and the escrow
                reverts with ``TermsMismatch()``.
        """
        ...

    def get_warrant(self, warrant_id: str) -> dict[str, Any] | None:
        """Read a warrant. ``None`` — not an exception — when it does not exist."""
        ...

    def list_warrants(self, query: dict[str, Any]) -> dict[str, Any]:
        """List an agent's warrants with aggregate statistics."""
        ...
